### ISSUE TITLE
Problem with using "navigateBack" with nested routes. Admin Dashboard Sidebar setup.

### ISSUE THREAD

**Issue by eXifreXi** [Milad-Akarie/auto_route_library#1207]
Hello there,

so I'm trying to get something similar to an Admin Dashboard working with Flutter (really not fun so far..) and I'm currently trying my luck with auto_route. I set up a simple test project to see if this is even working out for me. The app should sooner or later run on Desktop, Web, and Mobile, but for now I'm only interested in Desktop and Web.

**I'm testing this only on Desktop at the moment!**

Based on the docs I'm can use `navigate` to move between routes and can then use `navigateBack` to move backwards, similar to going back in web history.

My AppRouter currently looks like this:

```
@MaterialAutoRouter(replaceInRouteName: 'Page,Route', routes: <AutoRoute>[
  CustomRoute(
    path: 'welcome',
    page: WelcomePage,
    initial: true,
  ),
  CustomRoute(path: 'dashboard', page: DashboardPage, guards: [
    AuthGuard
  ], children: [
    CustomRoute(
      path: 'home',
      page: HomePage,
      initial: true,
    ),
    CustomRoute(
        path: 'clients',
        page: EmptyRouterPage,
        name: 'ClientsRoute',
        children: [
          CustomRoute(
            path: '',
            page: ClientsListPage,
          ),
          CustomRoute(
            path: ':userId',
            page: ClientsViewPage,
          )
        ]),
  ])
])
class AppRouter extends _$AppRouter {
  AppRouter({required AuthGuard authGuard}) : super(authGuard: authGuard);
}
```

_The AuthGuard can be ignored. Just a test and that's currently always calling `resolver.next(true);`._

**Here is how the Pages are set up:**

**WelcomePage** has a simple Login button, which calls: `AutoRouter.of(context).navigate(const DashboardRoute());`.

**DashboardPage** is a bit more complex. I use the builder of the AutoRouter Widget to create a Scaffold with a Row. That Row ultimately has a Column and then an expanded `child` (being the AutoRouter Widget if I understood correctly). That setup results in a sidebar on the left and an AutoRouter on the right.

Now the Sidebar has 4 Buttons. One button that calls `AutoRouter.of(context).navigate(const HomeRoute());`, HomeRoute being the default Route of the DashboardRoute, one button that calls `AutoRouter.of(context).navigate(const ClientsRoute());`, one that calls `context.router.navigateBack();` and one that calls `AutoRouter.of(context).navigate(const WelcomeRoute());`, which ultimately would send you back to the WelcomePage as a sort of Logout.

Looks like this in the end:

![image](https://user-images.githubusercontent.com/8512512/187033344-b107a6df-a066-49e0-980c-2ac0668f38da.png)

So far so good. The ClientsRoute, as seen in the AppRouter I posted, has a List and a View. List shows all Clients, while View shows one selected Client. The ClientListPage has a button that calls `AutoRouter.of(context).navigate(const ClientsViewRoute());`. At this point, I'm aware that the ClientsRoute has its own instance of AutoRouter, but I don't know what that changes for my `navigateBack` (see below).

This works so far, I can navigate from WelcomePage to DashboardPage. HomePage is initially selected. I can Logout. I can go Back from the HomePage to the WelcomePage via the Back button. I can navigate to the ClientsListPage as well as further to the ClientsViewPage. And I can navigate Back to the HomePage too.

The thing that causes me a headache is the Back button when navigating a few times in a row.
If I navigate WelcomePage->HomePage->ClientsListPage and press Back, I go back to the HomePage. That's fair. But if I press it again, I go back to the ClientsListPage and that does not make sense to me.
History-wise I should go back to the WelcomePage, but pressing Back just flip flops between ClientsListPage and HomePage.

If I go WelcomePage->HomePage->CientsListPage->ClientsViewPage and I press Back a few times, it goes
ClientsViewPage->HomePage (Skips ClientsListPage? Is that related to the second AutoRouter?)->ClientsListPage->HomePage->ClientsListPage....

Why is that? Did I set up something wrong or is this expected behavior on Desktop? Is there an example on a working setup for this kind of project? I have a hard time setting up Navigation for an app like this in Flutter ):

**Comment by eXifreXi** [comment]
Small extra comment, cause I'm still testing around while waiting for help.

I added another "Clients" setup, called "Projects". So same thing with an EmptyRouterPage, a List and a View, and I added another button to the DashboardPage. I wanted to test if it's maybe FlipFlopping in general, but that doesn't seem so.

Here is how it behaves:

WelcomePage->HomePage->ClientsPage(ClientsListPage)->ProjectsPage(ProjectsListPage)

Then using Back Button:

ProjectsPage->ClientsPage->HomePage->ClientsPage->HomePage->...

As well as:

WelcomePage->HomePage->ClientsPage->ProjectsPage->ClientsPage->ProjectsPage

Then using Back Button:

ProjectsPage->ClientsPage->HomePage->ClientsPage->HomePage->...

So what exactly is NavigateBack even doing on windows? It is that it brings me back to the previous location.
Am I missing some additional "pop" call or so?

**Comment by eXifreXi** [comment]
If I use push instead of navigate and pop instead of navigateBack, it works as expected.
But do I break anything Web-wise if I use push and pop? Also is there a limit to how many times one can push?
There is also the issue that push can be called when already being on that route, which I could test against I guess.

**Comment by eXifreXi** [comment]
So I tested navigateBack on web, and there it works fine. Basically exactly like the back button on the WebBrowser.
I also went ahead and checked how that even all works and the native implementation (which Desktop uses I guess), does:

```
  @override
  void back() {
    if (canNavigateBack) {
      _entries.removeLast();
      router.navigateAll([_entries.last.route]);
    }
  }
```

My random guess is that I'm just handling different nested routers wrong. I tried printing the current name, URL, etc. but I don't really understand what's going on. I did read up on the "AutoLeadingButton" now, and using that one seems to work, but I don't quite get why. It calls `scope.controller.popTop`. Why does that work, but navigateBack does not?

**Comment by Milad-Akarie** [comment]
@eXifreXi I don't know where to start, 
Multi stacks navigation is a complicated matter, typical web platforms have a linear navigation stack "/parent/child" child here does not have it's own stack it's just a child of the parent route, Mobile on the other hand has to have a sub-stack for each child route 'child' could be a tab that has it's own router (stack).

to explain how popTop works let's assume the following hierarchy
   
```
- /page1       -------->  a page inside root router
- /page2       -------->  a page inside root router which is also a router of sub-routes
       - /sub1     -------->  a page inside 'page2Router'
       - /sub2     -------->  a page inside 'page2Router'
- /page3       -------->  a page inside root router
```
now  calling router.popTop() from anywhere in the widgets three will pop '/page3' route, meaning it's gonna find the topMost router in hierarchy and call pop on it.
now our stack should look like this
```
- /page1       -------->  a page inside root router
- /page2       -------->  a page inside root router which is also a router of sub-routes
       - /sub1     -------->  a page inside 'page2Router'
       - /sub2     -------->  a page inside 'page2Router'
```
router.popTop()  from here will pop '/sub2' route which is a child route of page2Router. 
the logic is simple ... if the topMost route a sub-router? call sub-router.pop() else  pop the route, and that works recursively.

NativeHistory is an attempt by me to create a something like browser History it might be a little buggy still, but I want to clarify that it's not as simple as storing path locations  because when working with native arguments are passed round instead of path/query parameters which are not easy to embed into the path.
so basically it stores the same path segments that would report to browser on web as route objects instead of string locations so users args are kept. 

if you can give me simple steps to reproduce the the navigate back issue I might be able to fix it.

**Comment by eXifreXi** [comment]
``@Milad-Akarie  Thanks for the answer. I'm not sure I can repro this with simple steps. 
You can see the Router setup in the OP.

HomePage, and ClientsViewPage just has a centered text.
ClientsListPage has a centered text and a button that calls `AutoRouter.of(context).navigate(const ClientsViewRoute());`.
WelcomePage has a centered text and a button that calls `AutoRouter.of(context).navigate(const DashboardRoute());`.
DashboardPage is the only more complex one:

```
@override
  Widget build(BuildContext context) {
    return AutoRouter(
      builder: (context, child) {
        return Scaffold(
          body: Row(
            children: [
              SizedBox(
                width: 250,
                child: Container(
                    color: Colors.grey,
                    child: Column(
                      children: [
                        Text(currentRoute),
                        const Padding(
                            padding: EdgeInsets.only(top: 10, bottom: 10)),
                        ElevatedButton(
                            onPressed: () {
                              context.router.navigate(const HomeRoute());
                            },
                            child: const Text('Home')),
                        const Padding(
                            padding: EdgeInsets.only(top: 10, bottom: 10)),
                        ElevatedButton(
                            onPressed: () {
                              context.router.navigate(const ClientsRoute());
                            },
                            child: const Text('Clients')),
                        const Padding(
                            padding: EdgeInsets.only(top: 10, bottom: 10)),
                        ElevatedButton(
                            onPressed: () {
                              context.router.navigate(const ProjectsRoute());
                            },
                            child: const Text('Projects')),
                        const Padding(
                            padding: EdgeInsets.only(top: 10, bottom: 10)),
                        ElevatedButton(
                            onPressed: () {
                              context.router.navigateBack();
                            },
                            child: const Text('Back')),
                        const Padding(
                            padding: EdgeInsets.only(top: 10, bottom: 10)),
                        ElevatedButton(
                            onPressed: () {
                              context.router.navigate(const WelcomeRoute());
                            },
                            child: const Text('Logout')),
                      ],
                    )),
              ),
              Expanded(
                // nested routes will be rendered here
                child: child,
              )
            ],
          ),
        );
      },
    );
```

Basically creating the image from OP. I don't have the actual code from yesterday at hand at the moment, but that should be all. What I did to get stuck in the HomePage->ClientsListPage loop is to first click on the button in the WelcomePage to go to the DashboardPage. From here on I clicked on the Clients button on the left and then on the button that is part of the ClientsListPage, ultimately ending up on ClientsViewPage.

WelcomePage(initial) -Buttonpress-> DashboardPage -automatically-> HomePage(initial) -Buttonpress-> ClientsListPage -Buttonpress-> ClientsViewPage

From here on I click on the Back Button that calls `context.router.navigateBack();` (see above code).

I think that was all. Not sure if that example can be slimmed down or if this specific setup causes it.

**Comment by ericmartineau** [comment]
I haven't read through all the samples people posted, but I did run into an issue when the router has no entries - the `pop` operation terminates instead of bubbling up to the parent.  
```
auto_route_navigator.dart:63
/// This line means that there is no Navigator rendered
final navigator = widget.router.hasEntries
        ? Navigator(...) : widget.placeholder(...);
```

Then in `routing_controller.dart:825`: 
```
    final NavigatorState? navigator = _navigatorKey.currentState;
    if (navigator == null) return SynchronousFuture<bool>(false);
```
Since there's no `Navigator` widget, the `currentState` is null.  In my case, I want the  `_parent` controller to `pop` instead of returning false here.

Of course, I may be doing something wrong as well.
